src/gui/qgsmapcanvas.cpp

```diff
@@ -795,6 +799,116 @@ void QgsMapCanvas::clearTemporalCache()
+          displayPrecision = 8;
+        else
+          displayPrecision = 9;
+      }
+      catch ( QgsCsException & )
+      {
+        displayPrecision = crs.mapUnits() == QgsUnitTypes::DistanceDegrees ? 5 : 3;
+      }
+
+      QAction *copyCoordinateAction = new QAction( QStringLiteral( "%3 (%1, %2)" ).arg(
+            QString::number( transformedPoint.x(), 'f', displayPrecision ),
+            QString::number( transformedPoint.y(), 'f', displayPrecision ),
+            identifier ), mMenu );
+
+      connect( copyCoordinateAction, &QAction::triggered, this, [displayPrecision, transformedPoint]
+      {
+        QClipboard *clipboard = QApplication::clipboard();
+
+        const QString coordinates = QString::number( transformedPoint.x(), 'f', displayPrecision ) + ',' + QString::number( transformedPoint.y(), 'f', displayPrecision );
+
+        //if we are on x11 system put text into selection ready for middle button pasting
+        if ( clipboard->supportsSelection() )
+        {
+          clipboard->setText( coordinates, QClipboard::Selection );
+        }
+        clipboard->setText( coordinates, QClipboard::Clipboard );
+
+      } );
+      copyCoordinateMenu->addAction( copyCoordinateAction );
+    }
+    catch ( QgsCsException & )
+    {
+
+    }
+  };
+
+  addCoordinateFormat( tr( "Map CRS — %1" ).arg( mSettings.destinationCrs().userFriendlyIdentifier( QgsCoordinateReferenceSystem::ShortString ) ), mSettings.destinationCrs() );
+  if ( mSettings.destinationCrs() != QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) )
+    addCoordinateFormat( tr( "WGS84" ), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) );
+
+  QgsSettings settings;
+  const QString customCrsString = settings.value( QStringLiteral( "qgis/custom_coordinate_crs" ) ).toString();
+  if ( !customCrsString.isEmpty() )
+  {
+    QgsCoordinateReferenceSystem customCrs( customCrsString );
+    if ( customCrs != mSettings.destinationCrs() && customCrs != QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) )
+    {
+      addCoordinateFormat( customCrs.userFriendlyIdentifier( QgsCoordinateReferenceSystem::ShortString ), customCrs );
+    }
+  }
+  copyCoordinateMenu->addSeparator();
+  QAction *setCustomCrsAction = new QAction( QStringLiteral( "Set Custom CRS…" ), mMenu );
+  connect( setCustomCrsAction, &QAction::triggered, this, [ = ]
+  {
+    QgsProjectionSelectionDialog selector( this );
+    selector.setCrs( QgsCoordinateReferenceSystem( customCrsString ) );
+    if ( selector.exec() )
+    {
+      QgsSettings().setValue( QStringLiteral( "qgis/custom_coordinate_crs" ), selector.crs().authid().isEmpty() ? selector.crs().toWkt( QgsCoordinateReferenceSystem::WKT_PREFERRED ) : selector.crs().authid() );
+    }
+  } );
+  copyCoordinateMenu->addAction( setCustomCrsAction );
+
+  mMenu->addMenu( copyCoordinateMenu );
+
+  if ( mMapTool )
+    mMapTool->populateContextMenu( mMenu );
+
+  mMenu->exec( event->globalPos() );
```

Comment:
@nyalldawson I wonder if you see any inconvenient to make this QMenu alive only it this scope? It could be convenient to embed QObject with this QMenu, and to delete this object after QMenu::exec().

The difficulty is maybe to bring the menu in the lambda...